src/v/wasm/parser/parser.cc

```diff
@@ -0,0 +1,567 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "wasm/parser/parser.h"
+
+#include "base/units.h"
+#include "bytes/iobuf_parser.h"
+#include "strings/utf8.h"
+#include "wasm/parser/leb128.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/coroutine/maybe_yield.hh>
+#include <seastar/util/variant_utils.hh>
+
+#include <absl/algorithm/container.h>
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+#include <sys/types.h>
+
+#include <algorithm>
+#include <array>
+#include <cstdint>
+#include <iterator>
+#include <ranges>
+#include <stdexcept>
+#include <variant>
+
+namespace wasm {
+
+namespace {
+std::string_view to_string(const wasm::val_type& vt) {
+    switch (vt) {
+    case wasm::val_type::i32:
+        return "i32";
+    case wasm::val_type::i64:
+        return "i64";
+    case wasm::val_type::f32:
+        return "f32";
+    case wasm::val_type::f64:
+        return "f64";
+    case wasm::val_type::v128:
+        return "v128";
+    case wasm::val_type::funcref:
+        return "funcref";
+    case wasm::val_type::externref:
+        return "externref";
+    }
+    return "unknown";
+}
+} // namespace
+
+std::ostream& operator<<(std::ostream& os, const function_signature& sig) {
+    fmt::print(
+      os,
+      "{{params:[{}],results:[{}]}}",
+      fmt::join(sig.parameters | std::views::transform(to_string), ","),
+      fmt::join(sig.results | std::views::transform(to_string), ","));
```

Comment:
I'm not sure if we use changes at all, but man was this easy to do - I don't know if this will break in the open source build...

Reply:
Yeah, it's pretty awesome. 

> don't know if this will break in the open source build

Good call out. I thought I saw a discussion recently about turning off open source builds on PRs...did anything come of that?

Reply:
They run nightly. I'll just keep this and watch and see - if the OSS build fails I'll pick up the pieces later.

Reply:
I am also not sure about the toolchain upgrade, I think fedora uses 1.72, which I believe will still compile wasmtime